metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -139,15 +147,47 @@ class CheckpointingService[F[_]: Sync: Applicative, N] private (
       ledgerStorage.put(ledger) >>
         viewStateStorage.setLastExecutedBlockHash(header.hash)
     } >>
-      ledgerTree.set(Map(header.hash -> (ledger, header))) >>
+      ledgerTree.set(Map(header.hash -> (ledger, header, None))) >>
       lastExecutedHeader.set(header)
   }
 
-  private def constructCheckpoint(
-      ledger: Ledger,
+  private def constructCheckpointCert(
+      commitHeader: Block.Header,
       commitQC: QuorumCertificate[CheckpointingAgreement]
-  ): F[Option[CheckpointCertificate]] =
-    ??? //TODO: PM-3110
+  ): F[Option[CheckpointCertificate]] = {
+
+    def traverseCommitPath(
+        ledgers: LedgerTree,
+        from: Block.Hash,
+        path: NonEmptyList[Block.Header]
+    ): Option[CheckpointCertificate] =
+      ledgers.get(from) match {
+        case Some((_, _, Some(cp))) =>
+          CheckpointCertificate(path, cp.checkpoint, cp.proof, commitQC).some
+        case Some((_, h, None)) =>
+          traverseCommitPath(ledgers, h.parentHash, path :+ h)
```

Comment:
Maybe it would make sense to reverse the order in `CheckpointCertificate` so `head` is where the checkpoint is and `last` is what the commit is about, so blocks are in natural order?

I thought the order from commit to checkpoint would facilitate checking, but don't have a strong opinion. 

I would be nice to prepend instead of append though.

Reply:
I think I would prefer the natural order. I don't think the order should really affect the complexity of checking the parent links.